v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/GenericRecordTypeConvertor.java

```diff
@@ -0,0 +1,385 @@
+ */
+public class GenericRecordTypeConvertor {
+  private static final Logger LOG = LoggerFactory.getLogger(GenericRecordTypeConvertor.class);
+
+  private final ISchemaMapper schemaMapper;
+
+  private final String namespace;
+
+  public GenericRecordTypeConvertor(ISchemaMapper schemaMapper, String namespace) {
+    this.schemaMapper = schemaMapper;
+    this.namespace = namespace;
+  }
+
+  /**
+   * This method takes in a generic record and returns a map between the Spanner column name and the
+   * corresponding Spanner column value. This handles the data conversion logic from a GenericRecord
+   * field to a spanner Value.
+   */
+  public Map<String, Value> transformChangeEvent(GenericRecord record, String srcTableName) {
+    Map<String, Value> result = new HashMap<>();
+    String spannerTableName = schemaMapper.getSpannerTableName(namespace, srcTableName);
+    List<String> spannerColNames = schemaMapper.getSpannerColumns(namespace, spannerTableName);
+    for (String spannerColName : spannerColNames) {
+      /**
+       * TODO: Handle columns that will not exist at source - synth id - shard id - multi-column
+       * transformations - auto-gen keys - Default columns - generated columns
+       */
+      String srcColName =
+          schemaMapper.getSourceColumnName(namespace, spannerTableName, spannerColName);
+      Type spannerColumnType =
+          schemaMapper.getSpannerColumnType(namespace, spannerTableName, spannerColName);
+      Value value =
+          getSpannerValue(
+              record.get(srcColName),
+              record.getSchema().getField(srcColName).schema(),
+              srcColName,
+              spannerColumnType);
+      result.put(spannerColName, value);
+    }
+    return result;
+  }
+
+  /** Extract the field value from Generic Record and try to convert it to @spannerType. */
+  Value getSpannerValue(
+      Object recordValue, Schema fieldSchema, String recordColName, Type spannerType) {
+    // Logical and record types should be converted to string.
+    if (fieldSchema.getLogicalType() != null) {
+      recordValue = handleLogicalFieldType(recordColName, recordValue, fieldSchema);
+    } else if (fieldSchema.getType().equals(Schema.Type.RECORD)) {
+      // Get the avro field of type record from the whole record.
+      recordValue = handleRecordFieldType(recordColName, (GenericRecord) recordValue, fieldSchema);
+    }
+    Dialect dialect = schemaMapper.getDialect();
+    if (AvroToValueMapper.convertorMap().get(dialect).containsKey(spannerType)) {
+      return AvroToValueMapper.convertorMap()
+          .get(dialect)
+          .get(spannerType)
+          .apply(recordValue, fieldSchema);
+    } else {
+      throw new IllegalArgumentException(
+          "Found unsupported Spanner column type("
+              + spannerType.getCode()
+              + ") for column "
+              + recordColName);
+    }
+  }
+
+  static Boolean avroFieldToBoolean(Object recordValue, Schema fieldSchema) {
```

Comment:
Should you be moving all of these methods to `AvroToValueMapper`? 

That class takes in Avro stuff and spews out Value stuff by design.